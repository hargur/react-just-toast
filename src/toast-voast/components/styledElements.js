// Importing npm modules
import styled, { keyframes } from 'styled-components';
// Importing constants
import { POSITION_DIMENSIONS } from '../toast.constants';
// Importing styling
import '../style.css';

const toastAnimation = keyframes`
from {
    width: 0px;
  }

  to {
    width: 100%;
  }`;

/* A function to pass the correct values of top, bottom, left, right position based on prop value  position */
const checkPositionValue = (position, totalHeight) =>
    typeof position !== 'undefined'
        ? typeof totalHeight !== 'undefined'
            ? `${position + totalHeight}px`
            : `${position}px`
        : 'auto';

/* Styling main div for the toast */
export const StyledDiv = styled.div`
    color: ${props => props.textColor};
    text-align: ${props => props.textAlignment};
    display: flex;
    justify-content: center;
    align-items: center;
    line-height: 14px;
    box-sizing: border-box;
    font-size: 11px;
    background: ${props => props.bgColor};
    position: fixed;
    width: ${props => (props.visible && '100%') || 0};
    overflow: hidden;
    min-height: ${props => `${props.height}px`};
    top: ${props => checkPositionValue(POSITION_DIMENSIONS[props.position].top, props.totalHeight)};
    left: ${props => checkPositionValue(POSITION_DIMENSIONS[props.position].left)};
    right: ${props => checkPositionValue(POSITION_DIMENSIONS[props.position].right)};
    bottom: ${props => checkPositionValue(POSITION_DIMENSIONS[props.position].bottom, props.totalHeight)};
    box-shadow: 0 3px 5px rgba(0, 0, 0, 0.6);
    border-radius: 2px;
    transition: all 0.5s;
    z-index: 999;
    -webkit-animation-name: ${toastAnimation}; /* Safari 4.0 - 8.0 */
    -webkit-animation-duration: 0.3s;
    -webkit-animation-iteration-count: 1;
    animation-duration: 0.3s;
    animation-iteration-count: 1;
    animation-name: ${toastAnimation};
`;

/* Styling buttons for the toast */
export const StyledButton = styled.button`
    background: none;
    padding: ${props => `${props.padding}px 27px ${props.padding}px ${props.padding}px`};
    border: none;
    outline: none;
    color: ${props => props.textColor};
    width: 100%;
`;

/* Styling anchors for link/navigation for the toast */
export const StyledAnchor = styled.a`
    display: block;
    color: ${props => props.textColor};
    text-decoration: none;
    padding: ${props => `${props.padding}px 27px ${props.padding}px ${props.padding}px`};
    width: 100%;
`;

/* Styling anchors for link/navigation for the toast */
export const StyledPara = styled.p`
    margin: 0
`;

/* Styling inner div when no links */
export const StyledInnerDiv = styled.div`
    color: ${props => props.textColor};
    padding: ${props => `${props.padding}px 27px ${props.padding}px ${props.padding}px`};
    width: 100%;
`;
