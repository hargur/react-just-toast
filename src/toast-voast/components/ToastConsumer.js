// Importing constants
import { TYPES, TYPE_TEXT_COLOR, TOAST_PROPERTIES } from '../toast.constants';
// Importing npm modules
import React, { useContext } from 'react';
import { ThemeProvider } from '../index';
import ToastContainer from './ToastContainer';

const toastConfig = {};

const ToastConsumer = parentProps => {
    const context = useContext(ThemeProvider);
    const mapOverTypes = type =>
        (toastConfig[type] = (message, options = {}) => {
            options.type = type;
            options.id = options.id || Math.floor(Math.random() * 10000);
            options.bgColor = options.bgColor || (type === 'default' ? parentProps.bgColor : TYPE_TEXT_COLOR[type]);

        
            TOAST_PROPERTIES.forEach(
                optionValue => (options[optionValue] = options[optionValue] || parentProps[optionValue])
            );
            const obj = context.state.toastElements;

            // Total calculative height from where the toast should appear
            options.totalHeight =
                obj.length > 0
                    ?
                    obj.reduce(
                        (total, toast) =>
                            total +
                              7 +
                              (toast.options.height > 2 * toast.options.padding
                                  ? toast.options.height
                                  : 2 * toast.options.padding),
                        0
                    )
                    : 0;

            // Checking whether the id exists or not and appending accordingly
            const idExistsAlready =
                obj.length > 0
                    ?
                    obj.reduce((exists, toast) => (exists || toast.options.id === options.id ? true : false), false)
                    : false;

            if (!idExistsAlready) {
                obj.push({ message, options });
                context.setValues(obj);
            }
        });

    /* Maintaining functions for calling toastConfig */

    TYPES.map(mapOverTypes);

    // Closing all the toasts present
    toastConfig.closeAll = () => context.setValues([]);

    /* Closing a particular toast with the Id mentoned ; If id isn't present then nothing happens */
    toastConfig.close = id => {
        if (typeof id === 'number' || typeof id === 'string') {
        
            const mapForIdSearch = context.state.toastElements.map(toast => toast.options.id);
            const indexForId = mapForIdSearch.indexOf(id);

            if (indexForId > -1) {
                const { height, padding } = context.state.toastElements[indexForId].options;

                // Closing with particular id
            
                let obj = context.state.toastElements.map((toast, index) => {
                    // To shift up/down while the other toasts are deleted
                    if (index > indexForId) {
                        toast.options.totalHeight =
                            toast.options.totalHeight - 7 - (height > 2 * padding ? height : 2 * padding);
                    }
                    return toast;
                });
            
                obj = obj.filter(item => item.options.id !== id);
                context.setValues(obj);
            }
        }
    };

    return (
        <ThemeProvider.Consumer>
            {context =>
                context.state.toastElements &&
            
                context.state.toastElements.map((toast, index) => (
                    <ToastContainer
                        message={toast.message}
                        options={toast.options}
                        index={index}
                        key={toast.options.id}
                    />
                ))
            }
        </ThemeProvider.Consumer>
    );
};

export default ToastConsumer;
export { toastConfig };
