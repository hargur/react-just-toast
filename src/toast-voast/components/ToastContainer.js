// Importing npm modules
import React, { useState, useEffect } from 'react';
import {Link,
  withRouter
} from "react-router-dom";
import PropTypes from 'prop-types';

// Components
import { toastConfig } from './ToastConsumer';
import { StyledAnchor, StyledButton, StyledDiv, StyledInnerDiv, StyledPara } from './styledElements';

// Importing styling
import '../style.css';
import { IMAGE_ALT_TEXT } from '../toast.constants';

const ToastContainer = props => {
    const {
        history,
        message,
        index,
        options: {
            textColor,
            position,
            height,
            bgColor,
            padding,
            type,
            totalHeight,
            textAlignment,
            closeButtonImageSrc,
            actionHref,
            actionPrefetch,
            onClickAction,
            actionLinkAs,
            hideOnRouteChange,
            innerChildren,
            timeout,
            id
        }
    } = props;

    const [animateIn, setAnimateIn] = useState(true);

    useEffect(() => {
        // To close automatically if timeout given
        timeout &&
            setTimeout(() => {
                setAnimateIn(false);
            }, timeout);

        history.listen(() => {
            hideOnRouteChange && setAnimateIn(false);
        });
        // eslint-disable-next-line
    }, []);

    useEffect(
     () => {
        // Removing the element from main list of toasts
        if (!animateIn) {
            setTimeout(() => toastConfig.close(id), 500);
        }
      },
        // eslint-disable-next-line
        [animateIn]
    );

    return (
        <StyledDiv
            visible={animateIn}
            textColor={textColor}
            position={position}
            bgColor={bgColor}
            height={height}
            type={type}
            index={index}
            textAlignment={textAlignment}
            totalHeight={totalHeight}
            className={type}
        >
            {animateIn && (
                <img
                    src={closeButtonImageSrc}
                    width="11px"
                    height="11px"
                    alt={IMAGE_ALT_TEXT.CLOSE_TOAST_ICON}
                    title="close-toast"
                    className="close-button"
                    onClick={setAnimateIn.bind(this, false)}
                />
            )}
            {animateIn && !actionHref && !onClickAction && (
                <StyledInnerDiv padding={padding} textColor={textColor}>
                    {innerChildren ? innerChildren : <StyledPara>{message}</StyledPara>}
                </StyledInnerDiv>
            )}
            {actionHref ? (
                <Link prefetch={actionPrefetch} href={actionHref} as={actionLinkAs}>
                    <StyledAnchor padding={padding} textColor={textColor} title={actionHref}>
                        {animateIn && (innerChildren ? innerChildren : <StyledPara>{message}</StyledPara>)}
                    </StyledAnchor>
                </Link>
            ) : (
                onClickAction && (
                    <StyledButton
                        title={`notification-${type}`}
                        onClick={onClickAction}
                        padding={padding}
                        textColor={textColor}
                    >
                        {animateIn && (innerChildren ? innerChildren : <StyledPara>{message}</StyledPara>)}
                    </StyledButton>
                )
            )}
        </StyledDiv>
    );
};

ToastContainer.propTypes = {
    message: PropTypes.string,
    options: PropTypes.shape({
        textColor: PropTypes.string,
        position: PropTypes.string,
        height: PropTypes.number,
        bgColor: PropTypes.string,
        padding: PropTypes.number,
        type: PropTypes.string,
        totalHeight: PropTypes.number,
        textAlignment: PropTypes.string,
        closeButtonImageSrc: PropTypes.string,
        actionHref: PropTypes.string,
        actionPrefetch: PropTypes.bool,
        onClickAction: PropTypes.func,
        actionLinkAs: PropTypes.string,
        hideOnRouteChange: PropTypes.bool,
        innerChildren: PropTypes.string,
        timeout: PropTypes.number,
        id: PropTypes.oneOfType([
            PropTypes.string,
            PropTypes.number
        ])
    }),
    index: PropTypes.number
};

export default withRouter(ToastContainer);
