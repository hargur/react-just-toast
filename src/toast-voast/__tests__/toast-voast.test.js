import React from 'react';
import Adapter from 'enzyme-adapter-react-16';
import { shallow, mount, configure } from 'enzyme';

import toJson from 'enzyme-to-json';
//Tested component.
import Toast from '../index';
import ToastContainer from '../components/ToastContainer';
import { toastConfig } from '../components/ToastConsumer';
import { BrowserRouter } from 'react-router-dom';

configure({adapter: new Adapter()});

describe('Testing Main Toast Components', () => {
    it('Toast should render correctly with no props', () => {
        const component = shallow(<Toast isTestEnv />);
        expect(toJson(component)).toMatchSnapshot();
    });

    it('Toast container should render correctly', () => {
        const component = shallow(<ToastContainer message="error" options={{ textColor: '#fff' }} index="1" />);
        expect(toJson(component)).toMatchSnapshot();
    });
});

describe('Testing Different Toast Scenarios', () => {
    it('Testing toast props', async () => {
        const push = jest.fn();
        const component = mount(<BrowserRouter><Toast isTestEnv /></BrowserRouter>);
        component.setProps({ history: { push } });
        toastConfig.success('Success Case', { bgColor: '#000' });
        component.update();
        expect(getComputedStyle(component.find('div.success').getDOMNode()).getPropertyValue('background')).toBe('rgb(0, 0, 0)');
    });

    it('Testing whether toast is displayed', () => {
        const push = jest.fn();
        const component = mount(<BrowserRouter><Toast isTestEnv /></BrowserRouter>);
        component.setProps({ history: { push } });
        toastConfig.success('Success Case');
        component.update();
        expect(component.find('div.success').length).toBe(1);
    });

    it('Testing close all', () => {
        const push = jest.fn();
        const component = mount(<BrowserRouter><Toast isTestEnv /></BrowserRouter>);
        component.setProps({ history: { push } });
        toastConfig.error('Something went wrong');
        toastConfig.success('Success Case');
        toastConfig.info('Info');
        toastConfig.closeAll();
        component.update();
        expect(
            component.find('div.error').length &&
                component.find('div.success').length &&
                component.find('div.info').length
        ).toBe(0);
    });

    it('Testing close button with Timeout', () => {
        const push = jest.fn();
        const component = mount(<BrowserRouter><Toast isTestEnv timeout={3000} /></BrowserRouter>);
        component.setProps({ history: { push } });
        toastConfig.error('Something went wrong');
        toastConfig.success('Success Case');
        component.update();
        component.find('div.error img.close-button').simulate('click');
        component.update();
        setTimeout(() => {
            expect(component.find('div.error').length).toBe(0);
        }, 3000);
    });

    it('Testing closing with id', () => {
        const push = jest.fn();
        const component = mount(<BrowserRouter><Toast isTestEnv timeout={3000} /></BrowserRouter>);
        component.setProps({ history: { push } });
        toastConfig.error('Something went wrong', { id: '123' });
        toastConfig.success('Success Case');
        toastConfig.info('Info');
        toastConfig.close('123');
        component.update();
        setTimeout(() => {
            expect(component.find('div.error').length).toBe(0);
        }, 3000);
    });
});

