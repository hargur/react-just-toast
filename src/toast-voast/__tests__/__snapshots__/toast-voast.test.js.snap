// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Main Toast Components Toast container should render correctly 1`] = `
<ContextConsumer>
  <Component />
</ContextConsumer>
`;

exports[`Testing Main Toast Components Toast should render correctly with no props 1`] = `
<ContextProvider
  value={
    Object {
      "setValues": [Function],
      "state": Object {
        "toastElements": Array [],
      },
    }
  }
>
  <ToastConsumer
    actionPrefetch={true}
    bgColor="#333"
    closeButtonImageSrc="images/close.svg"
    height={30}
    hideOnRouteChange={true}
    isTestEnv={true}
    padding={10}
    position="topLeft"
    textAlignment="center"
    textColor="#ffffff"
    timeout={0}
  />
</ContextProvider>
`;
